### PR TITLE
Add a script to export mosaicGPT model with onnx

### PR DESCRIPTION
**Pull request by dskhudia**
This script takes a checkpoint, replaces flash/triton attn with torch attention and exports the model in onnx format. Checkpoint and exported model paths can be object store. 

You can specify the same config as training but some parameters (load_path) are required. 

Example usage:

- Local checkpoint + local export

`python scripts/export_for_inference.py yamls/mosaic_gpt/1b.yaml load_path=ep0-ba20-rank0.pt export_save_path=export_test/model.onnx` 

- To verify the exported model
`python scripts/export_for_inference.py yamls/mosaic_gpt/1b.yaml load_path=ep0-ba20-rank0.pt export_save_path=export_test/model.onnx verify_exported_model=True`  

- remote checkpoint + local export
`python scripts/export_for_inference.py yamls/mosaic_gpt/1b.yaml load_path=s3://my-bucket/my-folder/gpt-1b/checkpoints/latest-rank{rank}.pt export_save_path=model.onnx`

- remote checkpoint + remote export
`python scripts/export_for_inference.py yamls/mosaic_gpt/1b.yaml load_path=s3://my-bucket/my-folder/gpt-1b/checkpoints/latest-rank{rank}.pt export_save_path=s3://my-bucket/my-folder/gpt-1b/exported/model.onnx`

Please note that this is made to work with v0.0.3 release and we may have to update it with the current main. 

Sample script output:
Note: please ignore the message about missing keys in the checkpoint. This is due to the name of params being
different in flash/triton vs torch causal attention.
```
/workdisk/daya/mosaic-examples/examples/llm/scripts/export_for_inference.py:99: UserWarning: Changing init_device to cpu for export!!
  warnings.warn('Changing init_device to cpu for export!!')
/workdisk/daya/mosaic-examples/examples/llm/scripts/export_for_inference.py:110: UserWarning: Using a batch size of 1 for export!!
  warnings.warn(f'Using a batch size of {export_batch_size} for export!!')
Initializing model...
/workdisk/daya/mosaic-examples/examples/llm/src/models/layers/attention.py:278: UserWarning: While `attn_impl: triton` can be faster than `attn_impl: flash` it uses more memory. When training larger models this can trigger alloc retries which hurts performance. If encountered, we recommend using `attn_impl: flash`.
  warnings.warn(
You are using cfg.init_device='cpu', but you can also use cfg.init_device="meta" with Composer + FSDP for fast initialization.
Loading model weights...
                                                                                                                                                                                                                                Replacing triton attention with torch causal attention
You are using cfg.init_device='cpu', but you can also use cfg.init_device="meta" with Composer + FSDP for fast initialization.
                                                                                                                                                                                                                                Found these missing keys in the checkpoint: model.attn_mask, model.transformer.blocks.0.causal_attn.mhsa.in_proj_weight, model.transformer.blocks.0.causal_attn.mhsa.in_proj_bias, model.transformer.blocks.1.causal_attn.mhsa.in_proj_weight, model.transformer.blocks.1.causal_attn.mhsa.in_proj_bias
Found these unexpected keys in the checkpoint: model.transformer.blocks.0.causal_attn.mhsa.Wqkv.weight, model.transformer.blocks.0.causal_attn.mhsa.Wqkv.bias, model.transformer.blocks.1.causal_attn.mhsa.Wqkv.weight, model.transformer.blocks.1.causal_attn.mhsa.Wqkv.bias

model loading done ...
/workdisk/daya/mosaic-examples/examples/llm/src/models/mosaic_gpt.py:147: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  S <= self.cfg.max_seq_len
/workdisk/daya/mosaic-examples/examples/llm/src/models/layers/attention.py:368: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if key_padding_mask.bool().logical_not().any():
exported model ouptut matches with unexported model!!
```

Some future improvements:

1. Don't download model twice if it's remote store
2. Init twice on cpu can also be expensive for larger model. Add support for directly loading checkpoint into a `meta` model

### CODE REVIEW

File: examples/llm/scripts/export_for_inference.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 MosaicML Examples authors
```
Comment (dakinggg) [inline examples/llm/scripts/export_for_inference.py:1]: Could you add a description for this script (similar to the PR description) in the header here?

**Review by dakinggg — COMMENTED** [review]
Not very familiar with the onnx stuff, but looks good as far as I can tell. Could you also paste some proof of it working into the PR description for posterity?

**Comment by hanlint** [comment]
Could you also add some documentation on what this script does at the top of the script, and the motivation? Thanks for putting this together!

File: examples/llm/scripts/export_for_inference.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                            reproducibility)
+from omegaconf import OmegaConf as om
+
+from examples.llm import TorchCausalAttention
+from examples.llm.src.model_registry import COMPOSER_MODEL_REGISTRY
+
+
+def gen_random_batch(batch_size, cfg):
+    # generate input batch of random data
+    batch = {}
```
Comment (hanlint) [inline examples/llm/scripts/export_for_inference.py:23]: this can be simplified as:
```
batch  = {
    'input_ids':
    'attention_mask':
}
```

File: examples/llm/scripts/export_for_inference.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+def main(cfg):
+    # Filter deprecation warning from torch internal usage
+    warnings.filterwarnings(
+        action='ignore',
+        category=UserWarning,
+        message=
+        f'torch.distributed.*_base is a private function and will be deprecated.*',
+    )
+
+    reproducibility.seed_all(cfg.get('seed', 17))
```
Comment (hanlint) [inline examples/llm/scripts/export_for_inference.py:55]: using a default 17 seed reads like a magical number. Perhaps:
```
if cfg.get('seed', None) is not None:
    reproduciblity.seed_all(...)
```

File: examples/llm/scripts/export_for_inference.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    init_device = cfg.model.get('init_device', 'cpu')
+    if init_device == 'meta':
+        warnings.warn('Changing init_device to cpu for export!!')
+        cfg.model.init_device = 'cpu'
+
+    load_path = cfg.get('load_path', None)
+    if load_path is None:
+        print('Error: checkpoint load_path is required for exporting.')
```
Comment (hanlint) [inline examples/llm/scripts/export_for_inference.py:64]: This should raise a `ValueError`

File: examples/llm/scripts/export_for_inference.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        trainer.export_for_inference(
+            save_format='onnx',
+            save_path=local_save_path,
+            sample_input=sample_input,
+        )
+
+        if cfg.get('verify_exported_model', False):
+            test_input = gen_random_batch(export_batch_size, cfg)
+            with torch.no_grad():
+                orig_out = export_model(test_input)
```
Comment (hanlint) [inline examples/llm/scripts/export_for_inference.py:141]: Shouldn't this be:
```
orig_out = orig_model(test_input)
```

To cerify that replacing the flash attention with torch casual attention is math equivalent?

Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py:141]: We can't run the orig_model with flash/triton if this script runs on a cpu only system.

File: examples/llm/scripts/export_for_inference.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    else:
+        warnings.warn(f"Using a batch size of {export_batch_size} for export!!")
+
+    # Build Model
+    print("Initializing model...")
+    orig_model = build_composer_model(cfg.model)
+
+    # Loading checkpoint using Trainer
+    print("Loading model weights...")
+    trainer = Trainer(model=orig_model, load_path=load_path, load_weights_only=True)
```
[the remark this answers is not in the document]
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py]: This is needed for both if/else paths.

File: examples/llm/scripts/export_for_inference.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    if init_device == "meta":
+        warnings.warn("Changing init_device to cpu for export!!")
+        cfg.model.init_device = "cpu"
+
+    load_path = cfg.get("load_path", None)
+    if load_path is None:
+        print("Error: checkpoint load_path is required for exporting.")
+        sys.exit(-1)
+
+    export_batch_size = 1
```
[the remark this answers is not in the document]
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py:67]: Exported model only works for this batch size.

File: examples/llm/scripts/export_for_inference.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    reproducibility.seed_all(cfg.get("seed", 17))
+
+    init_device = cfg.model.get("init_device", "cpu")
+    if init_device == "meta":
+        warnings.warn("Changing init_device to cpu for export!!")
+        cfg.model.init_device = "cpu"
+
+    load_path = cfg.get("load_path", None)
+    if load_path is None:
+        print("Error: checkpoint load_path is required for exporting.")
```
[the remark this answers is not in the document]
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py]: Done

File: examples/llm/scripts/export_for_inference.py (hunk 1 above)
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py:1]: Done

File: examples/llm/scripts/export_for_inference.py (hunk 2 above)
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py]: done

File: examples/llm/scripts/export_for_inference.py (hunk 3 above)
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py]: Actually this is not needed at all. Removed it.

File: examples/llm/scripts/export_for_inference.py (hunk 4 above)
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py]: done

**Comment by dskhudia** [comment]
> Could you also paste some proof of it working into the PR description for posterity?

@dakinggg added sample output from a successful run of this script or did you request anything else?

**Comment by dakinggg** [comment]
> > Could you also paste some proof of it working into the PR description for posterity?
> 
> @dakinggg added sample output from a successful run of this script or did you request anything else?

Thats great, thanks!

File: examples/llm/scripts/export_for_inference.py (hunk 5 above)
Comment (hanlint) [reply examples/llm/scripts/export_for_inference.py:176]: Got it -- is it possible then to run this on a GPU system just to verify that the attention substitution process is indeed mathematically equivalent?

**Review by dakinggg — APPROVED** [review]
LGTM

File: examples/llm/scripts/export_for_inference.py (hunk 5 above)
Comment (dskhudia) [reply examples/llm/scripts/export_for_inference.py:176]: Yes. Just verified. They don't match (bf16 for the GPU run while fp32 for the cpu run) with the tolerance we have but look close enough. 

tensor([ 0.0977,  0.4785,  0.2363,  ..., -0.3965,  0.1172,  0.1182],
       device='cuda:0', dtype=torch.bfloat16)
[ 0.09721886  0.47875217  0.23697849 ... -0.3966748   0.11686482
  0.11804749]
